R/targets-utils.R

```diff
@@ -3,211 +3,105 @@
-        function(absent_fc) {
-          is_missing <- grepl(absent_fc, forecaster_ids)
-          params_only <- forecasters[is_missing]
-          mapply(c, params_only, id = forecaster_ids[is_missing])
-        }
-      ))
-    )
-  flat_missing <- unlist(grid_with_missing$missing_forecasters, recursive = FALSE)
-  unique_missing <- flat_missing[map_vec(flat_missing, \(x) length(x) > 0)] %>% unique()
-  return(unique_missing)
+forecaster_lookup <- function(forecaster_grid, pattern) {
+  forecaster_grid %>% filter(grepl(pattern, id))
 }
 
 #' Add a unique id based on the column contents
 #'
-#' Create a string of `n_adj` that is a hash of the parameters and append the
-#' `ahead` at the end.
+#' Create a string of `n_adj` that is a hash of the parameters.
 #'
-#' @param df the df to add a column to. everything should be convertable to a string
-#' @param n_adj the number of adjectives to use; default of 2.
+#' @param tib the tibble to add a column to. everything should be convertable to a string
+#' @param exclude a vector of column names to exclude from the hash
 #'
-#' @importFrom cli hash_animal
 #' @export
-add_id <- function(df, n_adj = 2) {
-  no_ahead <- df %>%
-    select(-ahead)
-  stringified <- no_ahead %>%
-    select(order(colnames(no_ahead))) %>%
+add_id <- function(tib, exclude = c()) {
+  tib %>%
+    select(all_of(order(colnames(.))) & !all_of(exclude)) %>%
     rowwise() %>%
-    mutate(id = paste(across(everything()), sep = "", collapse = ""), .keep = "none") %>%
-    mutate(id = hash_animal(id, n_adj = n_adj)$words) %>%
-    mutate(id = paste(id[1:n_adj], sep = "", collapse = "."))
-  df %<>%
+    mutate(id = get_single_id(across(everything()))) %>%
     ungroup() %>%
-    mutate(parent_id = stringified$id) %>%
-    rowwise() %>%
-    mutate(id = paste(parent_id, ahead, sep = ".", collapse = " ")) %>%
-    ungroup()
-  return(df)
+    relocate(id, .before = everything())
 }
 
-#' Generate an id from a simple list of parameters
+#' Generate a two-word id from a simple list of parameters.
+#'
+#' Assumes the list keys are sorted.
 #'
 #' @param param_list the list of parameters. must include `ahead` if `ahead = NULL`
-#' @param ahead the ahead to use.
 #'
-#' @inheritParams add_id
 #' @export
-single_id <- function(param_list, ahead = NULL, n_adj = 2) {
-  full_hash <- param_list[names(param_list) != "ahead"] %>%
-    .[order(names(.))] %>% # put in alphabetical order
-    lapply(function(x) if (length(x) > 1) list(x) else x) %>% # the tibble version needs vectors to actually be lists, so this is a conversion to make sure the strings are identical
-    paste(collapse = "") %>%
-    hash_animal(n_adj = n_adj)
-  single_string <- full_hash$words[[1]][1:n_adj] %>% paste(sep = ".", collapse = ".")
-  if (is.null(ahead)) {
-    full_name <- paste(single_string, param_list$ahead, sep = ".")
-  } else {
-    full_name <- paste(single_string, ahead, sep = ".")
-  }
-  return(full_name)
+get_single_id <- function(param_list) {
```

Comment:
So this no longer has guarantees about id order; given usage patterns, those should remain.

Reply:
yea, fair, i can pull the sorting in here.